TableView/TableView/DetailViewController.swift

```diff
@@ -14,16 +14,77 @@ class DetailViewController: UIViewController {
     @IBOutlet var imageView: UIImageView!
     @IBOutlet var nameLabel: UILabel!
     @IBOutlet var prepTime: UILabel!
+    @IBOutlet var cookedButton: UIButton!
+    @IBOutlet var ratingControl: UISegmentedControl!
 
-    var recipe: Recipe?
+    var recipeIndex: Int?
+    var recipes: [Recipe]?
+    var recipe: Recipe? {
+        guard let recipeIndex = recipeIndex else { return nil }
+        guard let recipes = recipes else { return nil }
+        return recipes[recipeIndex]
+    }
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        if let recipe = recipe {
-            navigationItem.title = recipe.name
-            imageView.image = UIImage(named: recipe.thumbnails)
-            nameLabel.text = recipe.name
-            prepTime.text = "Prep Time: " + recipe.prepTime
+        updateView()
+    }
+
+    func updateView() {
+        guard let recipe = recipe else { return }
+        navigationItem.title = recipe.name
+        imageView.image = UIImage(named: recipe.thumbnails)
+        nameLabel.text = recipe.name
+        prepTime.text = "Prep Time: " + recipe.prepTime
+        cookedButton.setTitle(recipe.cooked ? "Cooked" : "Not Cooked", for: .normal)
+        ratingControl.selectedSegmentIndex = (recipe.rating != nil) ? recipe.rating! : 0
+    }
+
+    @IBAction
+    func cookedAction() {
+        guard let recipe = recipe else { return }
+        guard let recipeIndex = recipeIndex else { return }
+        let updatedRecipe = Recipe(name: recipe.name, thumbnails: recipe.thumbnails, prepTime: recipe.prepTime, rating: recipe.rating, cooked: !recipe.cooked)
+        self.recipes![recipeIndex] = updatedRecipe
+        self.saveRecipes()
+        self.updateView()
+    }
+
+    @IBAction
+    func ratingAction() {
+        guard let recipe = recipe else { return }
+        guard let recipeIndex = recipeIndex else { return }
+        let updatedRecipe = Recipe(name: recipe.name, thumbnails: recipe.thumbnails, prepTime: recipe.prepTime, rating: ratingControl.selectedSegmentIndex, cooked: recipe.cooked)
+        self.recipes![recipeIndex] = updatedRecipe
+        self.saveRecipes()
+        self.updateView()
+    }
+
+    func saveRecipes() {
+        guard let recipes = recipes else { return }
+
+        // Show saving modal
+        let savingView = UIViewController()
+        let savingLabel = UILabel()
+        savingLabel.text = "Saving..."
+        savingView.view.addSubview(savingLabel)
+        self.present(savingView, animated: true, completion: nil)
+
+        // Save recipe data to defaults
+        DispatchQueue.global().async {
+            do {
+                let recipeJSON = try JSONEncoder().encode(recipes)
+                UserDefaults.init(suiteName: "Recipes Defaults")?.set(recipeJSON, forKey: "Recipes")
```

Comment:
Instead of propagating this change via the user defaults, consider using delegates or notifications. This would make it more efficient.